### ISSUE TITLE
Use SnpLinAlg for sparse models as well

### ISSUE THREAD

**Issue by biona001** [OpenMendel/MendelIHT.jl#46]
Currently, `xk` in `IHTVariable` (and `Xk` in `mIHTVariable`) is a n by k `Matrix{T}`. For highly polygenic traits like height, `k` could be really large (>10000), which for large sample sizes requires a lot of memory. 

For cross validation, we could recommend users to cross validate only one (or a few) `k` at once. However, for the long run, we should consider using `SnpLinAlg` for `xk`.

**Comment by biona001** [comment]
After extensive benchmark, it turns out that `copyto!(v.xk, @view(v.x[:, v.idx]))` at [this line](https://github.com/OpenMendel/MendelIHT.jl/blob/master/src/utilities.jl#L94) is very slow if `k` is large (here `v.x` is a `SnpLinAlg`). 

Maybe we need another `update_xb!` that does dense SnpLinAlg-vector multiplication when `k` is large. Here's some preliminary benchmarks

```julia
# Computes xb = x * b for sparse b by xk=x[:, idx], bk=b[idx], then computing xb = xk * bk
function update_xb!(xb, x, xk, b, bk, idx)
    copyto!(xk, @view(x[:, idx]))
    copyto!(bk, view(b, idx))
    mul!(xb, xk, bk)
end

using MendelIHT, SnpArrays, LinearAlgebra, BenchmarkTools, Random
BLAS.set_num_threads(1)
x = SnpArray("/scratch/users/bbchu/ukb_prs/fastphase/ukb_data/ukb.QC.train0.8.bed")
xla = SnpLinAlg{Float64}(x, center=true, scale=true, impute=true)
n, p = size(x)
b = rand(p)
xb = zeros(n)
k = 30000
gk = zeros(k)
xk = zeros(n, k)
idx = sort!(shuffle!(collect(1:p))[1:k])

# n = 10000, p = 500000 (test on mac, 8 Julia threads)
@btime SnpArrays.mul!($xb, $xla, $b); # 2.168 s (18223 allocations: 1.07 MiB
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 100:   8.250 ms (0 allocations: 0 bytes)
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 1000:  80.442 ms (0 allocations: 0 bytes)
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 10000: 759.911 ms (0 allocations: 0 bytes)

# n = 256075, p = 591513 (test on slurm, 16 Julia threads)
@btime SnpArrays.mul!($xb, $xla, $b); # 19.969 s (437089 allocations: 25.30 MiB)
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 100:   100.945 ms (0 allocations: 0 bytes)
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 1000:  1.069 s (0 allocations: 0 bytes) 
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 10000: 11.646 s (0 allocations: 0 bytes)
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 30000: 37.406 s (0 allocations: 0 bytes)
```
I wonder if there's some threshold involving `n`, `p`, `k`, where doing dense SnpLinAlg-vector multiplication will definitely be better than copying. 

Another option is to parallelize the [copyto!](https://github.com/OpenMendel/SnpArrays.jl/blob/master/src/linalg_direct.jl#L845) function.

**Comment by biona001** [comment]
```julia
using MendelIHT, SnpArrays, LinearAlgebra, BenchmarkTools, Random
BLAS.set_num_threads(1)

# Computes xb = x * b for sparse b by xk=x[:, idx], bk=b[idx], then computing xb = xk * bk
function update_xb!(xb, x, xk, b, bk, idx)
    copyto!(xk, @view(x[:, idx]))
    copyto!(bk, view(b, idx))
    mul!(xb, xk, bk)
end

function copyto_threaded!(
    v::AbstractVecOrMat{T}, 
    s::AbstractSnpLinAlg
    ) where T <: AbstractFloat
    m, n = size(s, 1), size(s, 2)
    Threads.@threads for j in 1:n
        @simd for i in 1:m
            @inbounds v[i, j] = s[i, j]
        end
    end
    return v
end

function update_xb_threaded2!(xb, x, xk, b, bk, idx)
    copyto_threaded2!(xk, @view(x[:, idx]))
    copyto!(bk, view(b, idx))
    mul!(xb, xk, bk)
end

function copyto_threaded2!(
    v::AbstractVecOrMat{T}, 
    s::AbstractSnpLinAlg
    ) where T <: AbstractFloat
    m, n = size(s, 1), size(s, 2)
    @sync for j in 1:n
        @inbounds for i in 1:m
            Threads.@spawn v[i, j] = s[i, j]
        end
    end
    return v
end

function update_xb_threaded!(xb, x, xk, b, bk, idx)
    copyto_threaded!(xk, @view(x[:, idx]))
    copyto!(bk, view(b, idx))
    mul!(xb, xk, bk)
end

# n, p = 10000, 500000
# x = simulate_random_snparray(undef, n, p)
x = SnpArray("/scratch/users/bbchu/ukb_prs/fastphase/ukb_data/ukb.QC.train0.8.bed")
n, p = size(x)
xla = SnpLinAlg{Float64}(x, center=true, scale=true, impute=true)
b = rand(p)
xb = zeros(n)
k = 100
gk = zeros(k)
xk = zeros(n, k)
idx = sort!(shuffle!(collect(1:p))[1:k])
b[setdiff(1:p, idx)] .= 0

# check answer
# xb1, xb2, xb3, xb4 = zeros(n), zeros(n), zeros(n), zeros(n)
# SnpArrays.mul!(xb1, xla, b);
# update_xb!(xb2, xla, xk, b, gk, idx);
# update_xb_threaded!(xb3, xla, xk, b, gk, idx);
# update_xb_threaded2!(xb4, xla, xk, b, gk, idx);
# [xb1 xb2 xb3 xb4]

# n = 10000, p = 500000 (test on mac, 8 Julia threads)
@btime SnpArrays.mul!($xb, $xla, $b); # 1.863 s (17730 allocations: 1.01 MiB)
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 100:   7.162 ms (0 allocations: 0 bytes)
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 1000:  66.174 ms (0 allocations: 0 bytes)
@btime update_xb!($xb, $xla, $xk, $b, $gk, $idx); # k = 10000: 1.165 s (0 allocations: 0 bytes)
@btime update_xb_threaded!($xb, $xla, $xk, $b, $gk, $idx); # k = 100:   1.300 ms (41 allocations: 3.73 KiB)
@btime update_xb_threaded!($xb, $xla, $xk, $b, $gk, $idx); # k = 1000:  13.736 ms (41 allocations: 3.73 KiB)
@btime update_xb_threaded!($xb, $xla, $xk, $b, $gk, $idx); # k = 10000: 276.264 ms (41 allocations: 3.73 KiB)
@btime update_xb_threaded2!($xb, $xla, $xk, $b, $gk, $idx); # k = 100:   2.701 s (7000024 allocations: 589.62 MiB)
@btime update_xb_threaded2!($xb, $xla, $xk, $b, $gk, $idx); # k = 1000:  
@btime update_xb_threaded2!($xb, $xla, $xk, $b, $gk, $idx); # k = 10000: 

```

**Comment by biona001** [comment]
Closed by #47
